programs/locker/src/util/token2022.rs

```diff
@@ -0,0 +1,250 @@
+    amount: u64,
+    token_mint: &InterfaceAccount<Mint>,
+) -> Result<u64> {
+    let mint_info = token_mint.to_account_info();
+    if *mint_info.owner == Token::id() {
+        return Ok(amount);
+    }
+
+    let token_mint_data = mint_info.try_borrow_data()?;
+    let token_mint_unpacked =
+        StateWithExtensions::<spl_token_2022::state::Mint>::unpack(&token_mint_data)?;
+    let actual_amount: u64 =
+        if let Ok(transfer_fee_config) = token_mint_unpacked.get_extension::<TransferFeeConfig>() {
+            let transfer_fee = transfer_fee_config.get_epoch_fee(Clock::get()?.epoch);
+            calculate_pre_fee_amount(transfer_fee, amount)
+                .ok_or(LockerError::TransferFeeCalculationFailure)?
+        } else {
+            amount
+        };
+
+    Ok(actual_amount)
+}
+
+// Memo Extension support
+pub fn is_transfer_memo_required(token_account: &InterfaceAccount<TokenAccount>) -> Result<bool> {
+    let token_account_info = token_account.to_account_info();
+    if *token_account_info.owner == Token::id() {
+        return Ok(false);
+    }
+
+    let token_account_data = token_account_info.try_borrow_data()?;
+    let token_account_unpacked =
+        StateWithExtensions::<spl_token_2022::state::Account>::unpack(&token_account_data)?;
+    let extension =
+        token_account_unpacked.get_extension::<extension::memo_transfer::MemoTransfer>();
+
+    return if let Ok(memo_transfer) = extension {
+        Ok(memo_transfer.require_incoming_transfer_memos.into())
+    } else {
+        Ok(false)
+    };
+}
+
+pub fn calculate_pre_fee_amount(transfer_fee: &TransferFee, post_fee_amount: u64) -> Option<u64> {
+    if post_fee_amount == 0 {
+        return Some(0);
+    }
+    let maximum_fee = u64::from(transfer_fee.maximum_fee);
+    let transfer_fee_basis_points = u16::from(transfer_fee.transfer_fee_basis_points) as u128;
+    if transfer_fee_basis_points == 0 {
+        Some(post_fee_amount)
+    } else if transfer_fee_basis_points == ONE_IN_BASIS_POINTS {
+        Some(maximum_fee.checked_add(post_fee_amount)?)
+    } else {
+        let numerator = (post_fee_amount as u128).checked_mul(ONE_IN_BASIS_POINTS)?;
+        let denominator = ONE_IN_BASIS_POINTS.checked_sub(transfer_fee_basis_points)?;
+        // let raw_pre_fee_amount = ceil_div(numerator, denominator)?;
+        let raw_pre_fee_amount = numerator
+            .checked_add(ONE_IN_BASIS_POINTS)?
+            .checked_sub(1)?
+            .checked_div(denominator)?;
+
+        if raw_pre_fee_amount.checked_sub(post_fee_amount as u128)? >= maximum_fee as u128 {
+            post_fee_amount.checked_add(maximum_fee)
+        } else {
+            // should return `None` if `pre_fee_amount` overflows
+            u64::try_from(raw_pre_fee_amount).ok()
+        }
+    }
+}
+
+/// Calculate the fee that would produce the given output
+pub fn calculate_inverse_fee(transfer_fee: &TransferFee, post_fee_amount: u64) -> Option<u64> {
```

Comment:
We dont use this function, so pls remove this, and rewrite test case

Reply:
updated